providers/implementations/macs/hmac_prov.c

```diff
@@ -141,19 +141,31 @@ static size_t hmac_size(void *vmacctx)
     return HMAC_size(macctx->ctx);
 }
 
-static int hmac_init(void *vmacctx)
+static int hmac_init(void *vmacctx, const unsigned char *key,
+                     size_t keylen, const OSSL_PARAM params[])
 {
     struct hmac_data_st *macctx = vmacctx;
     const EVP_MD *digest;
     int rv = 1;
 
-    if (!ossl_prov_is_running())
+    if (!ossl_prov_is_running() || !hmac_set_ctx_params(macctx, params))
         return 0;
 
+    if (key != NULL) {
```

Comment:
I dont think I like this duping of code..

Reply:
Me neither.  @paulidale, please make a separate function that treats key and key length appropriately, and that gets called both from here and from the set_params function.

This goes for the same sort of code copying in the other implementations

Reply:
I see that this was already mentioned elsewhere

Reply:
Done it for all the MACs.